nfsv4/client.c

```diff
@@ -21,50 +21,83 @@ void fill_default_user(nfs4_properties p)
         (__s)?(__s)->error:0;                                          \
     })
 
+int nfs4_error_num(nfs4 n)
+{
+    return n->nfs_error_num;
+}
+
 char *nfs4_error_string(nfs4 n)
 {
     return n->error_string->contents;
 }
 
-// should return the number of bytes read, can be short
+// Read from f into dest and return the number of bytes read or -1 upon error
+// the error is recorded in the nfs4 
 int nfs4_pread(nfs4_file f, void *dest, bytes offset, bytes len)
 {
     u64 total = 0;
 
     while (1) {
         rpc r = file_rpc(f);
         u64 transferred = push_read(r, offset,  dest + total, len - total, &f->latest_sid);
-        total += transferred;
-        offset += transferred;
-        if (total < len) {
+        status s;
+        if (total + transferred < len) {
             // reestablish connection here
-            api_check(r->c, rpc_send(r));
+            s = rpc_send(r);
         } else {
             // block on all the preceeding(?) reads
-            return api_check(f->c, transact(r));
+            s = transact(r);
+        }
+        if (nfs4_is_error(s)) {
+            if (total == 0) {
+                f->c->error_string = s->description;
+                f->c->nfs_error_num = s->error;
+                return -1;
+            } else {
+                return total;
+            }
+        }
+        total += transferred;
+        offset += transferred;
+        if (total >= len) {
+            return total;
         }
     }
 }
 
+// Write from source into f and return bytes written or -1 upon error.
 int nfs4_pwrite(nfs4_file f, void *source, bytes offset, bytes length)
 {
+    u64 total = 0;
+
     // not in the asynch case
     buffer b = alloca_wrap_buffer(source, length);
-    while (buffer_length(b)) {
+    while (1) {
         rpc r = file_rpc(f);
-        offset += push_write(r, offset, b, &f->latest_sid);
-
+        u64 transferred = push_write(r, offset, b, &f->latest_sid);
+        status s;
         if (buffer_length(b)) {
-          api_check(r->c, rpc_send(r));
+            s = rpc_send(r);
         } else {
-          return api_check(f->c, transact(r));
+            s = transact(r);
+        }
+        if (nfs4_is_error(s)) {
+            if (total == 0) {
```

Comment:
This conditional statement looks like it may swallow errors if total > 0. In what cases does that occur? Are we sure that is the correct behavior?

Reply:
Hmm so I looked back again to the [man pages](https://linux.die.net/man/2/write)

> The number of bytes written may be less than count if, for example, there is insufficient space on the underlying physical medium, or the RLIMIT_FSIZE resource limit is encountered (see setrlimit(2)), or the call was interrupted by a signal handler after having written less than count bytes. (See also pipe(7).) 

The page suggests that returning partial writes is okay for certain errors - right now we're catching all errors. 

I think the slightly weird part is that we could have part 1 of the write succeed and part 2 fail. A return value of -1 is ambiguous as to how much progress was made. I guess users should expect the ambiguity when an error occurs - I think for this PR it's best to change to always return -1 upon error regardless of total.

A follow up task to catch the edge case errors mentioned in man can fix the rest - I also think we don't encounter those very often.

Reply:
Ok, I agree. Yes, you really can't be sure about what happened in a write when there's a failure.

Reply:
Implemented